xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1125 @@
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+}
+
+// TestTwoBalancersSameType tests the scenario where there is a graceful switch
+// load balancer setup with a current and pending Load Balancer of the same
+// type. Any ClientConn update should be forwarded to the current if no pending,
+// and the only the pending if the graceful switch balancer contains both a
+// current and pending. The pending Load Balancer should also swap into current
+// whenever it updates with a connectivity state other than CONNECTING.
+func (s) TestTwoBalancersSameType(t *testing.T) {
+	tcc, gsb := setup(t)
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to either the Current or Pending Load
+	// Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	gsb.SwitchTo(mockBalancerBuilder1{})
+	gsb.UpdateClientConnState(ccs)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+
+	// The current balancer reporting READY should cause this state
+	// to be forwarded to the Client Conn.
+	gsb.balancerCurrent.Balancer.(*mockBalancer).updateState(balancer.State{
+		ConnectivityState: connectivity.Ready,
+		Picker:            &neverErrPicker{},
+	})
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
+	case state := <-tcc.NewStateCh:
+		if state != connectivity.Ready {
+			t.Fatal("wanted connectivity state ready")
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
+	case picker := <-tcc.NewPickerCh:
+		// Should receive a never err picker.
+		_, err := picker.Pick(balancer.PickInfo{})
+		if err != nil {
+			t.Fatalf("Graceful Switch Balancer should have sent a never error picker, which is a picker that never errors, instead received err %v", err)
+		}
+	}
+
+	// An explicit call to switchTo, even if the same type, should cause the
+	// balancer to build a new balancer for pending.
+	gsb.SwitchTo(mockBalancerBuilder1{})
+	if gsb.balancerPending == nil {
+		t.Fatal("balancerPending was not built out when another SwitchTo() call should've triggered the pending balancer to build")
+	}
+
+	// A Client Conn update received should be forwarded to the new pending LB
+	// policy, and not the current one.
+	gsb.UpdateClientConnState(ccs)
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err == nil {
```

Comment:
Make it explicit that a context with a short deadline is being used:

```
	sCtx, sCancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
	defer sCancel()
	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(sCtx, ccs); err == nil { ... }
```

Reply:
Done across whole file.